Diz.Core/model/snes/Data.cs

```diff
@@ -0,0 +1,471 @@
+            if (!romWord.HasValue || !rb3Null.HasValue)
+                return null;
+            
+            return romWord + (rb3Null << 16);
+        }
+        
+        public int? GetRomDoubleWord(int offset)
+        {
+            if (offset + 3 >= GetRomSize())
+                return null;
+
+            var romLong = GetRomLong(offset);
+            var rb4Null = GetRomByte(offset + 3);
+            if (!romLong.HasValue || !rb4Null.HasValue)
+                return null;
+            
+            return romLong + (rb4Null << 24);
+        }
+        
+        public int GetIntermediateAddressOrPointer(int offset)
+        {
+            switch (GetFlag(offset))
+            {
+                case FlagType.Unreached:
+                case FlagType.Opcode:
+                    return GetIntermediateAddress(offset, true);
+                case FlagType.Pointer16Bit:
+                    int bank = GetDataBank(offset);
+                    var romWord = GetRomWord(offset);
+                    if (!romWord.HasValue)
+                        return -1;
+                    
+                    return (bank << 16) | (int)romWord;
+                case FlagType.Pointer24Bit:
+                case FlagType.Pointer32Bit:
+                    var romLong = GetRomLong(offset);
+                    if (!romLong.HasValue)
+                        return -1;
+                    
+                    return (int)romLong;
+            }
+            return -1;
+        }
+        
+        // NOTE: technically not always correct. banks wrap around so, theoretically we should check what operation
+        // we're doing and wrap to the beginning of the bank. for now.... just glossing over it, bigger fish to fry.
+        // "past me" apologizes to 'future you' for this if you got hung up here.
+        //
+        // returns null if out of bounds
+        public byte? GetNextRomByte(int pcOffset)
+        {
+            return pcOffset + 1 >= 0 && pcOffset + 1 < RomByteSource.Bytes.Count
+                ? RomByteSource.Bytes[pcOffset + 1].Byte
+                : null;
+        }
+
+        public int GetBankSize()
+        {
+            return RomUtil.GetBankSize(RomMapMode);
+        }
+
+        public ByteEntry BuildFlatByteEntryForSnes(int snesAddress)
+        {
+            return SnesAddressSpace?.BuildFlatByteEntryFor(snesAddress);
+        }
+
+        public int ConvertSnesToPc(int snesAddress)
+        {
+            return RomUtil.ConvertSnesToPc(snesAddress, RomMapMode, GetRomSize());
+        }
+        
+        public int ConvertPCtoSnes(int pcOffset)
+        {
+            return RomUtil.ConvertPCtoSnes(pcOffset, RomMapMode, RomSpeed);
+        }
+
+        private Cpu CpuAt(int offset) => new CpuDispatcher().Cpu(this, offset); 
+
+        public int Step(int offset, bool branch, bool force, int prevOffset)
+        {
+            return CpuAt(offset).Step(this, offset, branch, force, prevOffset);
+        }
+
```

Comment:
where possible, we're probably at the point where a lot of the Cpu and Mark operations don't make a ton of sense to be in here anymore, probably move them into their own classes / extension methods / etc.